Module health node manager

tommyp1ckles opened the pull request:
This PR seeks to provide modular health status checking on the "node manager" component of the Agent. Specifically the goals are:

## Intentions of this PR:

### Bubble-up all errors in Linux node handler code

Node Manager related code relies heavily on logging (or just ignoring) errors far down the stack, instead of bubbling them up.

The result of this is:
* Logged errors have little context on the root event causing them to be invoked.
* It's hard to trace failure from a particular event relating to all reconcile procedures that need to run.
* It's hard to definitively to say if reconciliation for a Node Event was successfully.

The changes in this PR seek to address this by handling and bubbling up all relevant errors that may be encountered while doing Linux Node update reconciliation logic.

As well, this PR seeks to **avoid** any changes to the control-flow logic of the reconciliation by _not_ immediately returning at places where errors may occur.
Rather, this follows a pattern of combining lists of errors into relevant/scoped multi errors (using the multierr library). All code will execute as before, with the difference being that errors are handled up the stack.

### Use Hive's Modular Health Reporter to report status of node manager.

This is a first attempt at trying to manage health status for a module. 

#### Follow up work:
* Investigate Reconciliation: Break Linux node logic into atomic and idempotent tasks that can be executed and retried (+ reported on more easily).

tommyp1ckles committed:
Add Name() function to NodeHandler interface to identify handlers.

The NodeHandler interface is used to chain handlers to be executed when Node updates occur.
We want to be able to better identify these handlers when they fail so we
collect and report on errors we can provide more context on where the
error occurred.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
node/manager: add health reporter and track errors from node manager

Inject cell.HealthReporter into module.

As well, add LRU error set to track module errors relating to
Update/Create/Delete events.

When Update/Create/Delete handlers are run by the node manager, the resulting error is now captured and if not nil added to the error set.

The periodic backgroundSync loop, which runs ever 1m will aggregate all errors and update the healthReporter that the module is degraded if there are any existing errors.

The error set is keyed using a "topic" that is a tuple of:

* Event type
* Node name
* Handler name

Thus, errors are cleared from the error set selectively by what kind of
error occurred, and will be cleared individually if the same type of
event/node/handler eventually succeeds.

This will provide a reasonable initial implementation of health checking
node manager.

We may run into situations where because errors must be cleared by their "topic" tuple - they may never actually be cleared.
The initial approach is to eventually begin evicting such errors,
eventually we may want to look at a ttl based approach.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: nodeUpdate aggregate errors for enable IPSec.

enableIPSec run when invoking updateNode(...), which is called from the
Update/Add/Validate node handler interface implementations for the
datapath linux node handler (linuxNodeHandler).

This refactors enableIPSec and all underlying functions to return an accumulated error upon failures.
IPSec enable exection will continue even if error is collected as to not affect control flow.

Top level enableIPSec error is returned wrapped to scope what part of the updateNode procedure failed.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: aggregate errors for node route update.

updateOrRemoveNodeRoutes() attempts to reconcile a diff between old and new routes, depending on routing mode.
Currently procedure tasks to apply the ip route change via netlink can fail and will not have their error logged or returned.

This updates the updateOrRemoveNodeRoutes code paths such that all errors are bubbled up and returned without affecting execution flow. Existing logging is preserved.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: add comment explaining neighbor error handling.

Neighbor handling is a bit tricky as it's done async, possibly in
parallel with other neighbor adds.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: aggregate errors for subnet ipsec.

Aggregate and return all errors encountered while performing ipsec: route procedure.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux: aggregate NodeDeleted event handler errors.

Similar to node update handler, node deleted needs to tear down some
host configuration state during a node delete.

This aggregates all errors returned in the NodeDelete procedure, and
returns error to node manager.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux: make time since last ping more readable.

time.Since(...) is easier to read/parse than using subtract.
Switch code to use this approach - this does not change behavior.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
datapath/linux/node: fix fmt.Errorf linting issues.

Use wrap directives instead of string for correct error
handling and fix capitalized error messages.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
hive/hivetest: add mock health reporter.

A mock implementation of cell.HealthReporter will be useful in the
future for testing as more modules depend on being passed a reporter.

This implementation simply keeps the last msg/err/stopped state.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

ti-mo commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -517,15 +519,16 @@ func ipSecReplacePolicyOut(src, dst *net.IPNet, tmplSrc, tmplDst net.IP, nodeID
 
-func ipsecDeleteXfrmState(nodeID uint16) {
+func ipsecDeleteXfrmState(nodeID uint16) error {
+	var acc error
```
This variable is never assigned to?

ti-mo commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -21,10 +21,12 @@ import (
 	"github.com/fsnotify/fsnotify"
 	"github.com/sirupsen/logrus"
 	"github.com/vishvananda/netlink"
+	"go.uber.org/multierr"
```
Go 1.20 introduced `errors.Join()`, maybe use that instead?

ti-mo commented on pkg/errs/__debug_bin:
What is this file for?

ti-mo reviewed (changes requested):
Right off the bat I have some concerns around the pervasive use of multierr and the introduction of `pkg/errs`, so I'm flagging this early before reviewing the whole thing. Perhaps the changes to errors can be isolated from the modularity work? That will make the PR more manageable to review.

If we're going to deviate from the canonical Go error handling style, I'd like this to be discussed in the open, e.g. through a CFP, and make it part of a larger plan to convert the whole codebase. We need strong reasons for doing so, and it's not a decision we should take lightly, as it makes code look very different due to how often a function like `errs.Into` appears. It directly affects the readablilty and familiarity of the code, so personally I'd stick to the idiomatic Go style.

I think errors.Join() could be used in the majority of cases here, which is not as big of a departure as multierr or a new package.

Another observation: I'm seeing a few occurrences of `acc` (what's in a name, btw?) declared and assigned to, but never returned. IMO the pattern of declaring a separate `acc` (instead of an `err` that is `errors.Join()`ed into) is a code smell, as it splinters the source of truth. `errors.Join` doesn't really have this problem; it supports nil values and the resulting error value will remain nil if the inputs are all nil.

chancez commented on pkg/hubble/peer/handler.go:
```diff
@@ -37,6 +37,10 @@ func newHandler(withoutTLSInfo bool, addressPref serviceoption.AddressFamilyPref
 }
 
+func (h *handler) Name() string {
+	return "hubble-peers"
```
Nit: should this be singular? Package is singular, and we refer to it as the "hubble peer service" elsewhere.

chancez replied:
🤔 did this change?

tommyp1ckles committed:
linux/ipsec/ipsec_linux: return errors from ipsec/xfrm functions.

Previously many ip xfrm netlink requests would fail, and either be logged or
have the error unhandled.

This adds error handling in xfrm commands that are used by procedures in linux datapath updateNode.

Eventually, we'll want to correctly bubble up all meaningful errors in this package.
This only changes functionality that affects linux node datapath update/delete/create logic.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>
